src/binfhe/lib/lwe-pke.cpp

```diff
@@ -50,6 +50,58 @@ LWEPrivateKey LWEEncryptionScheme::KeyGen(usint size, const NativeInteger& modul
     return std::make_shared<LWEPrivateKeyImpl>(LWEPrivateKeyImpl(tug.GenerateVector(size, modulus)));
 }
 
+// size is the ring dimension N, modulus is the large Q used in RGSW encryption of bootstrapping.
+LWEKeyPair LWEEncryptionScheme::KeyGenPair(const std::shared_ptr<LWECryptoParams> params) const {
+    int size              = params->GetN();
+    NativeInteger modulus = params->GetQ();
+
+    // generate secret vector s of ring dimension N
+    auto skN = KeyGen(size, modulus);
+    auto pkN = PubKeyGen(params, skN);
+
+    auto lweKeyPair = LWEKeyPairImpl(pkN, skN);
+
+    // return the public key (A, v), private key sk pair
+    return std::make_shared<LWEKeyPairImpl>(lweKeyPair);
+}
+
+// size is the ring dimension N, modulus is the large Q used in RGSW encryption of bootstrapping.
+LWEPublicKey LWEEncryptionScheme::PubKeyGen(const std::shared_ptr<LWECryptoParams> params,
+                                            ConstLWEPrivateKey skN) const {
+    int size              = params->GetN();
+    NativeInteger modulus = params->GetQ();
+
+    DiscreteUniformGeneratorImpl<NativeVector> dug;
+    dug.SetModulus(modulus);
+    std::vector<NativeVector> A(size);
+
+    // generate random matrix A of dimension N x N
+    for (int i = 0; i < size; i++) {
+        NativeVector a = dug.GenerateVector(size);
+        A[i]           = std::move(a);
+    }
+
+    // generate error vector e
+    DiscreteGaussianGeneratorImpl<NativeVector> dgg;
+    NativeVector e = dgg.GenerateVector(size, modulus);
+
+    // compute v = As + e
+    NativeInteger mu = modulus.ComputeMu();
+    NativeVector v   = e;
+    NativeVector ske = skN->GetElement();
+    for (int j = 0; j < size; ++j) {
+        for (int i = 0; i < size; ++i) {
+            v[j] += A[i][j].ModMulFast(ske[i], modulus, mu);
```

Comment:
Should be able to use vector-level operations. They are faster than iterative over each integer in a vector.